RAC-406: Make a job stoppable (#1528)
akeneo/pim-docs: import_and_export_data/guides/create-connector.rst

        protected FindItemsToProcess $findItemsToProcess;
        protected JobStopper $jobStopper;

        public function __construct(
            FindItemsToProcess $findItemsToProcess,
            JobStopper $jobStopper
        ) {
            $this->jobStopper = $jobStopper;
            $this->findItemsToProcess = $findItemsToProcess;
        }

        public function setStepExecution(StepExecution $stepExecution): void
        {
            $this->stepExecution = $stepExecution;
        }

        public function execute(): void
        {
            $itemsToProcess = $this->FindItemsToProcess->find();
            foreach ($itemsToProcess as $i => $itemToProcess) {
                $itemToProcess->doSomeWork();

                // Check every 100 items if the process should be stopped
                if ($i % self::BATCH_SIZE === 0
                    && $this->jobStopper->isStopping($this->stepExecution)
                ) {
                    $this->jobStopper->stop($this->stepExecution);

                    return;
                }
            }
        }
    }

Track the progress of a job
---------------------------

> As of now, it is possible your users are not able to see the progress of your job in the UI.

Reviewer: If it is a 5.0 feature too, you need to say it explicitly I think (like you did to stop a job feature).

The following example shows a simple tasklet updating it's progress using the step execution.

.. code-block:: php
    :linenos:

    <?php
    declare(strict_types=1);

    namespace Acme\Bundle\StoppableJobBundle;

    class TrackableTasklet implements TaskletInterface, TrackableTaskletInterface
    {
        protected ?StepExecution $stepExecution = null;
        protected FindItemsToProcess $findItemsToProcess;
        protected JobRepositoryInterface $jobRepository;

        public function __construct(
            FindItemsToProcess $findItemsToProcess,
            JobRepositoryInterface $jobRepository
        ) {
            $this->findItemsToProcess = $findItemsToProcess;
            $this->jobRepository = $jobRepository;
        }

        public function setStepExecution(StepExecution $stepExecution): void
        {
            $this->stepExecution = $stepExecution;
        }

        public function execute(): void
        {
            // First, let's calculate the total items to process

>             $itemsToProcess = $this->FindItemsToProcess->find();

Reviewer 2: ```suggestion
            $itemsToProcess = $this->findItemsToProcess->find();
```

Now that we have our extension loaded in our form, we can add some logic into it, check how to `customize the UI`_.

.. _customize the UI: ../../design_pim

Make a job stoppable
--------------------

> As of now, it is possible your jobs won't be able to stop when the job is started.

Reviewer 3: Isn't weird to start by saying it does not work sometimes?

Author: because it depends on their custom job. If they use item steps it might work, if they use something else, it's possible it does not :/

Reviewer 4: What do you think about:

In Akeneo 5.0 we added the possibility to declare a Job as stoppable. But as some task may be not stoppable (important workflows that always need to finish, job depending on other steps, etc) we decided to make this new feature "opt in". In this documentation, we will explain you how to make your job stoppable.

Author: fixed

Reviewer: I agree with Julien, it's better like this.

To make this action available from the UI, we need to make sure of a few things:

1. You need to make sure your custom job implements ``Akeneo\Tool\Component\Batch\Job\StoppableJobInterface`` and set the last property of the constructor to "true". See the following job service definition example:

.. code-block:: yaml
    :linenos:

    pimee_catalog_rule.job.xlsx_product_import_with_rules:
        class: '%pim_connector.job.simple_job.class%'
        arguments:
            - '%pimee_catalog_rule.job_name.xlsx_product_import_with_rules%'
            - '@event_dispatcher'
            - '@akeneo_batch.job_repository'
            -
                - '@pim_connector.step.charset_validator'
                - '@pimee_catalog_rule.step.xlsx_product.import'
                - '@pimee_catalog_rule.step.xlsx_product.import_associations'
                - '@pimee_catalog_rule.step.xlsx_product.execute_rules'
            - true # <-- This property should be true if the job is stoppable
        tags:
            - { name: akeneo_batch.job, connector: '%pim_connector.connector_name.xlsx%', type: '%pim_connector.job.import_type%' }

> 2. If your job uses tasklets (by implementing ``TaskletInterface``) to execute your business logic. You can inject the service `"akeneo_batch.job.job_stopper"` directly in your custom tasklets and use it to indicate your job is stopping and then stops.

Reviewer 5: ```suggestion
2. If your job uses tasklets (by implementing ``TaskletInterface``) to execute your business logic. You can inject the service `"akeneo_batch.job.job_stopper"` directly in your custom tasklets and use the function `JobStopper::isStopping` to know if user asked to stop the job. If this function return true, you should use the function `JobStopper::stop` and exit your Tasklet.

 .. warning:: 
 
Make sure that for large set of items to process, you should not call `JobStopper::isStopped` function at each time you processed an item, but rather 100 by 100. 
```

Author: fixed

Now that we have our extension loaded in our form, we can add some logic into it, check how to `customize the UI`_.

.. _customize the UI: ../../design_pim

Make a job stoppable
--------------------

As of now, it is possible your jobs won't be able to stop when the job is started.

.. note::

>     If your job is based on the ``Akeneo\Tool\Component\Batch\Job\Job`` class, you only have to adapt your job service definition by setting the last property of the constructor to "true" (as shown in the example below).

Reviewer 4: ```suggestion
    If your job is based on the ``Akeneo\Tool\Component\Batch\Job\Job`` class, you have to adapt your job service definition by setting the last property of the constructor to "true" (as shown in the example below).
```

It's often advised to avoid "easy, only, simple" words as they may imply that something is easy or simple and could be frustrating if you are stuck

    pimee_catalog_rule.job.xlsx_product_import_with_rules:
        class: '%pim_connector.job.simple_job.class%'
        arguments:
            - '%pimee_catalog_rule.job_name.xlsx_product_import_with_rules%'
            - '@event_dispatcher'
            - '@akeneo_batch.job_repository'
            -
                - '@pim_connector.step.charset_validator'
                - '@pimee_catalog_rule.step.xlsx_product.import'
                - '@pimee_catalog_rule.step.xlsx_product.import_associations'
                - '@pimee_catalog_rule.step.xlsx_product.execute_rules'
            - true # <-- This property should be true if the job is stoppable
        tags:
            - { name: akeneo_batch.job, connector: '%pim_connector.connector_name.xlsx%', type: '%pim_connector.job.import_type%' }

2. If your job uses tasklets (by implementing ``TaskletInterface``) to execute your business logic. You can inject the service `"akeneo_batch.job.job_stopper"` directly in your custom tasklets and use the function `JobStopper::isStopping` to know if user asked to stop the job. If this function return true, you should use the function `JobStopper::stop` and exit your Tasklet.

 .. warning:: 
 

> Make sure that for large set of items to process, you should not call `JobStopper::isStopped` function at each time you processed an item, but rather 100 by 100. 

Reviewer 4: ```suggestion
Make sure to only call `JobStopper::isStopped` not too often as it will do a MySQL query. A good way of doing it could be to do it only every 100 loop tour. 
```

Author: fixed

Track the progress of a job
---------------------------

As of now, it is possible your users are not able to see the progress of your job in the UI.

.. note::

    If your job is composed of an ``Akeneo\Tool\Component\Batch\Step\ItemStep`` instance, as well as a reader implementing ``Akeneo\Tool\Component\Batch\Item\TrackableItemReaderInterface`` the tracking of your job should already be available in the UI.

> If your job uses a custom reader, make sure it implements ``Akeneo\Tool\Component\Batch\Item\TrackableItemReaderInterface`` and exposes the total of number that will be processed during the execution of the step.

Reviewer 4: ```suggestion
If your job uses a custom reader, make sure it implements ``Akeneo\Tool\Component\Batch\Item\TrackableItemReaderInterface`` and exposes the total of number of items that will be processed during the execution of the step.
```

If your job uses a custom tasklet, we need to make sure of a few additional things:

- your tasklet should implement the ``Akeneo\Tool\Component\Batch\Item\TrackableTaskletInterface`` interface
- at the very beginning of the execution of the tasklet, you need to provide the step execution with the total items your tasklet will process through the ``Akeneo\Tool\Component\Batch\Model\StepExecution::setTotalItems`` function.

> - during the process of the tasklet, you need to provide the step execution with the progression the tasklet by incrementing a counter through the ``Akeneo\Tool\Component\Batch\Model\StepExecution::incrementProcessedItems`` function.

Reviewer 4: maybe a small example would help here

Author: I agree! Let me write a php class here that does the job of a simple tasklet!

Author: done
